h2o-algos/src/main/java/hex/tree/ScoreBuildHistogram2.java

```diff
@@ -0,0 +1,354 @@
+      return res;
+    }
+
+
+    @Override
+    protected void map(int id){
+      cs = MemoryManager.malloc8d(_maxChunkSz);
+      ys = MemoryManager.malloc8d(_maxChunkSz);
+      ws = MemoryManager.malloc8d(_maxChunkSz);
+      // start computing
+      if(_unordered) {
+        for(DHistogram [] dhary:_lhcs)
+          for(DHistogram dh:dhary)
+            if(dh != null) dh.init();
+      } else if(!_shareHisto) {
+        for (int l = _leaf; l < _tree._len; l++) {
+          DTree.UndecidedNode udn = _tree.undecided(l);
+          DHistogram hs[] = _lhcs[l - _leaf];
+          int sCols[] = udn._scoreCols;
+          if (sCols != null) { // Sub-selecting just some columns?
+            for (int col : sCols) // For tracked cols
+              if (_colFrom <= col && col < _colTo) hs[col - _colFrom].init();
+          } else {                 // Else all columns
+            for (int j = 0; j < hs.length; j++) // For all columns
+              if (hs[j] != null)        // Tracking this column?
+                hs[j].init();
+          }
+        }
+      }
+      for(int i = _cidx.getAndIncrement(); i < _cids.length; i = _cidx.getAndIncrement())
+        computeChunk(i);
+    }
+
+    public void updateHistoUnordered(DHistogram[] hcs, int len, double[] ws, double[] cs, double[] ys, int [] nids){
+//      double minmax[] = new double[]{_min2,_maxIn};
+      // Gather all the data for this set of rows, for 1 column and 1 split/NID
+      // Gather min/max, wY and sum-squares.
+      for(int r = 0; r< len; ++r) {
+        double w = ws[r];
+        if (w == 0) continue;
+        int nid = nids[r];
+        if(nid < 0) continue; // decided row?
+        DHistogram dh = hcs[nid];
+        if(dh == null) continue;
+        dh.updateHisto(w, cs[r], ys[r]);
+      }
+      for(DHistogram dh:hcs)
+        if(dh != null)dh.reducePrecision();;
+    }
+
+    private void computeChunk(int id){
+      ScoreBuildHistogram.LocalHisto lh = _shareHisto?new ScoreBuildHistogram.LocalHisto(Math.max(_nbins,_nbins_cats)):null;
+      int cidx = _cids[id];
+      int [] nh = _nhs[id];
+      int [] rs = _rss[id];
+      int [] nnids = _unordered?_nnids[id]:null;
+      Chunk resChk = _fr2.vec(_workIdx).chunkForChunkIdx(cidx);
+      int len = resChk._len;
+      resChk.getDoubles(ys, 0, len);
+      if(_weightIdx != -1)
+        _fr2.vec(_weightIdx).chunkForChunkIdx(cidx).getDoubles(ws, 0, len);
+      else
+        Arrays.fill(ws,1);
+      final int hcslen = _lhcs.length;
+      for (int c = _colFrom; c < _colTo; c++) {
+        if(_unordered){
+          _fr2.vec(c).chunkForChunkIdx(cidx).getDoubles(cs, 0, len);
+          updateHistoUnordered(_lhcs[c-_colFrom],len,ws,cs,ys,nnids);
+          for(DHistogram[] dha:_lhcs)
+            for(DHistogram dh:dha)
+              if(dh != null) dh.reducePrecision();
```

Comment:
hmm, yeah, it is also called at the very end in postGlobal. I'll remove the extra calls. The unordered branch does not really work in all cases, maybe I should just remove it altogether.